File: src/2025-03-04-03-28-40--deepseek-chat-0304/go/exercises/practice/alphametics/alphametics.go
package alphametics

import (
	"errors"
	"strings"
	"unicode"
)

func Solve(puzzle string) (map[string]int, error) {
	parts := strings.Split(puzzle, "==")
	if len(parts) != 2 {
		return nil, errors.New("invalid puzzle format")
	}

	leftWords := strings.Fields(parts[0])
	rightWord := strings.TrimSpace(parts[1])

	// Collect all unique letters and identify leading letters
	letters := make(map[rune]bool)
	leadingLetters := make(map[rune]bool)
	for _, word := range append(leftWords, rightWord) {
		for i, char := range word {
			if unicode.IsLetter(char) {
				letters[char] = true
				if i == 0 {
					leadingLetters[char] = true
				}
			}
		}
	}

	uniqueLetters := make([]rune, 0, len(letters))
	for char := range letters {
		uniqueLetters = append(uniqueLetters, char)
	}

	// Try all permutations using backtracking
	mapping := make(map[string]int)
	usedDigits := make(map[int]bool)
	if solveRecursive(uniqueLetters, 0, mapping, usedDigits, leftWords, rightWord, leadingLetters) {
		return mapping, nil
	}

	return nil, errors.New("no solution found")
}

func solveRecursive(letters []rune, index int, mapping map[string]int, usedDigits map[int]bool,
	leftWords []string, rightWord string, leadingLetters map[rune]bool) bool {
	if index == len(letters) {
		return isValidMapping(leftWords, rightWord, mapping)
	}

	char := letters[index]
	for digit := 0; digit <= 9; digit++ {
		// Skip if digit is already used or if it's a leading zero
		if usedDigits[digit] || (digit == 0 && leadingLetters[char]) {
			continue
		}

		mapping[string(char)] = digit
		usedDigits[digit] = true

		if solveRecursive(letters, index+1, mapping, usedDigits, leftWords, rightWord, leadingLetters) {
			return true
		}

		delete(mapping, string(char))
		usedDigits[digit] = false
	}

	return false
}

func isValidMapping(leftWords []string, rightWord string, mapping map[string]int) bool {
	leftSum := 0
	for _, word := range leftWords {
		num := 0
		for _, char := range word {
			num = num*10 + mapping[string(char)]
		}
		leftSum += num
	}

	rightSum := 0
	for _, char := range rightWord {
		rightSum = rightSum*10 + mapping[string(char)]
	}

	return leftSum == rightSum
}

File: src/2025-03-04-03-28-40--deepseek-chat-0304/javascript/exercises/practice/alphametics/alphametics.js
function solveEquation(puzzle, letters, mapping) {
    const [left, right] = puzzle.split('=').map(s => s.trim());
    const leftWords = left.split('+').map(s => s.trim());
    
    let leftSum = 0;
    for (const word of leftWords) {
        let num = 0;
        for (const letter of word) {
            num = num * 10 + mapping[letter];
        }
        leftSum += num;
    }
    
    let rightSum = 0;
    for (const letter of right) {
        rightSum = rightSum * 10 + mapping[letter];
    }
    
    return leftSum === rightSum;
}

function assignDigits(letters, leadingLetters, digitsUsed, index, mapping, puzzle) {
    if (index === letters.length) {
        return solveEquation(puzzle, letters, mapping) ? {...mapping} : null;
    }
    
    const letter = letters[index];
    for (let digit = 0; digit <= 9; digit++) {
        if (digitsUsed[digit]) continue;
        if (digit === 0 && leadingLetters.has(letter)) continue;
        
        digitsUsed[digit] = true;
        mapping[letter] = digit;
        
        const result = assignDigits(letters, leadingLetters, digitsUsed, index + 1, mapping, puzzle);
        if (result) return result;
        
        digitsUsed[digit] = false;
    }
    
    return null;
}

export const solve = (puzzle) => {
    const letters = [...new Set(puzzle.replace(/[^A-Z]/g, ''))];
    const [left, right] = puzzle.split('=').map(s => s.trim());
    const leftWords = left.split('+').map(s => s.trim());
    
    const leadingLetters = new Set([
        ...leftWords.map(word => word[0]),
        right[0]
    ]);
    
    const digitsUsed = Array(10).fill(false);
    const mapping = {};
    
    return assignDigits(letters, leadingLetters, digitsUsed, 0, mapping, puzzle);
};
